src/tree/updater_quantile_hist.cc

```diff
@@ -620,7 +636,7 @@ void QuantileHistMaker::Builder<GradientSumT>::Update(
 template<typename GradientSumT>
 bool QuantileHistMaker::Builder<GradientSumT>::UpdatePredictionCache(
     const DMatrix* data,
-    HostDeviceVector<bst_float>* p_out_preds) {
+    HostDeviceVector<bst_float>* p_out_preds, const int gid, const int ngroup) {
```

Comment:
Can you get ```gid``` and ```ngroup``` from the Builder rather than pass them as arguments?
You have ```p_last_tree_ pointer``` . Probably, it's enough already or need to just add a few fields into builder.

In this case:
1) No needs to have ```UpdatePredictionCacheMulticlass``` function at all.
2) No changes in arguments of` ```UpdatePredictionCache```
3) No changes in basic classes like: ```GBTree``` and ```TreeUpdater```

Reply:
We have to provide `gid` at each `UpdatePredictionCache` call. Currently there is no possibility to get group id `gid` in specific Builder and even in `QuantileHistMaker`. It requires extensions for `QuantileHistMaker` with methods like `setGroupIndex()`  for example. But as `UpdatePredictionCache` call is polymorphic we can't avoid changes in basic  `TreeUpdater` class.